numpy/core/src/_simd/_simd_inc_convert.h

```diff
@@ -0,0 +1,206 @@
+#if !NPY_SIMD
+    #error "Not a standalone header, only works through 'simd.dispatch.c.src'"
+#endif
+
+/************************************
+ ** Protected Definitions
+ ************************************/
+static simd_data
+simd_scalar_from_obj(PyObject *obj, simd_data_type dtype)
+{
+    const simd_data_info *info = simd_data_getinfo(dtype);
+    assert(info->is_scalar && info->lane_size > 0);
+    simd_data data;
+    if (info->is_float) {
+        data.f64 = PyFloat_AsDouble(obj);
+        if (dtype == simd_data_f32){
+            data.f32 = (float)data.f64;
+        }
+    } else {
+        data.u64 = PyLong_AsUnsignedLongLongMask(obj);
+    }
+    return data;
+}
+
+static PyObject *
+simd_scalar_to_obj(simd_data data, simd_data_type dtype)
+{
+    const simd_data_info *info = simd_data_getinfo(dtype);
+    assert(info->is_scalar && info->lane_size > 0);
+    if (info->is_float) {
+        if (dtype == simd_data_f32) {
+            return PyFloat_FromDouble(data.f32);
+        }
+        return PyFloat_FromDouble(data.f64);
+    }
+    int leftb = (sizeof(npyv_lanetype_u64) - info->lane_size) * 8;
+    data.u64 <<= leftb;
+    if (info->is_signed) {
+        return PyLong_FromLongLong(data.s64 >> leftb);
+    }
+    return PyLong_FromUnsignedLongLong(data.u64 >> leftb);
+}
+
+static void *
+simd_sequence_new(Py_ssize_t len, simd_data_type dtype)
+{
+    const simd_data_info *info = simd_data_getinfo(dtype);
+    assert(info->is_sequence && info->lane_size > 0);
+
+    size_t size  = NPY_SIMD_WIDTH + sizeof(size_t) + sizeof(size_t*);
+           size += len * info->lane_size;
+
+    size_t *ptr = malloc(size);
+    if (ptr == NULL) {
+        return PyErr_NoMemory();
+    }
+    *(ptr++) = len;
+    size_t **a_ptr = (size_t**)(
+        ((size_t)ptr + NPY_SIMD_WIDTH) & ~(size_t)(NPY_SIMD_WIDTH-1)
+    );
+    a_ptr[-1] = ptr;
+    return a_ptr;
+}
+
+static size_t
+simd_sequence_len(const void *ptr)
+{
+    size_t *ptrz = ((size_t**)ptr)[-1];
+    return *(ptrz-1);
+}
+
+static void
+simd_sequence_free(void *ptr)
+{
+    size_t *ptrz = ((size_t**)ptr)[-1];
+    free(ptrz-1);
+}
```

Comment:
Could use a struct here to cleanup the pointer arithmetic
```suggestion
typedef struct {
    size_t len;
    void *ptr;
} simd_alloc_data;

static void *
simd_sequence_new(Py_ssize_t len, simd_data_type dtype)
{
    const simd_data_info *info = simd_data_getinfo(dtype);
    assert(info->is_sequence && info->lane_size > 0);

    size_t size = sizeof(simd_alloc_data) + len * info->lane_size + NPY_SIMD_WIDTH;

    void *ptr = malloc(size);
    if (ptr == NULL) {
        return PyErr_NoMemory();
    }
    // align the pointer
    simd_alloc_data *a_ptr = (simd_alloc_data *)(
        ((uintptr_t)ptr + sizeof(simd_alloc_data) + NPY_SIMD_WIDTH) & ~(uintptr_t)(NPY_SIMD_WIDTH-1)
    );
    a_ptr[-1].len = len;
    a_ptr[-1].ptr = ptr;
    return a_ptr;
}

static size_t
simd_sequence_len(void const *ptr)
{
    return ((simd_alloc_data const*)ptr)[-1].len;
}

static void
simd_sequence_free(void *ptr)
{
    free(((simd_alloc_data const*)ptr)[-1].ptr);
}
```

Reply:
looks far way better, thank you. just changed the prefix of `alloc_data` to `simd__` since I use two underscores for private
definitions.